### PR TITLE
Fix stack overflow

### LINKED ISSUE #654: Stack Overflow in SingleDocParser::HandleFlowMap()
Stack Overflow in SingleDocParser::HandleFlowMap()

**position：**[code](https://github.com/jbeder/yaml-cpp/blob/562aefc114938e388457e6a531ed7b54d9dc1b62/src/singledocparser.cpp#L275)

**[poc](https://github.com/jbeder/yaml-cpp/files/2714271/crash9.txt)**

**To reproduce:** `./parse < crash9.txt`

**gdb:**
`Program received signal SIGSEGV, Segmentation fault.`
`0x08081d5d in YAML::SingleDocParser::HandleFlowMap(YAML::EventHandler&) ()`

**ASAN:**
`ASAN:DEADLYSIGNAL`
`=================================================================
==6596==ERROR: AddressSanitizer: stack-overflow on address 0xbf09dee0 (pc 0x0814e423 bp 0xbf09e308 sp 0xbf09ded0 T0)`
`    #0 0x814e422  (/home/user/yaml-cpp/build/util/parse+0x814e422)`
`    #1 0x816fd48  (/home/user/yaml-cpp/build/util/parse+0x816fd48)`
`    #2 0x8190a72  (/home/user/yaml-cpp/build/util/parse+0x8190a72)`
`    #3 0x81d5061  (/home/user/yaml-cpp/build/util/parse+0x81d5061)`
`    #4 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`    #5 0x81d511a  (/home/user/yaml-cpp/build/util/parse+0x81d511a)`
`    #6 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`    #7 0x81d511a  (/home/user/yaml-cpp/build/util/parse+0x81d511a)`
`    #8 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`    #9 0x81d511a  (/home/user/yaml-cpp/build/util/parse+0x81d511a)`
`    #10 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`    ......                                                                                                         `
`    #247 0x81d511a  (/home/user/yaml-cpp/build/util/parse+0x81d511a)`
`    #248 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`    #249 0x81d511a  (/home/user/yaml-cpp/build/util/parse+0x81d511a)`
`    #250 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`SUMMARY: AddressSanitizer: stack-overflow (/home/user/yaml-cpp/build/util/parse+0x814e422) 
==6596==ABORTING`

### LINKED ISSUE #655: Stack Overflow in Scanner::EnsureTokensInQueue()
Stack Overflow in Scanner::EnsureTokensInQueue() 

**position：**[code](https://github.com/jbeder/yaml-cpp/blob/562aefc114938e388457e6a531ed7b54d9dc1b62/src/scanner.cpp#L48)

**[poc](https://github.com/jbeder/yaml-cpp/files/2714360/crash2.txt)**

**To reproduce:** `./parse < crash2.txt`

**gdb:**
`Program received signal SIGSEGV, Segmentation fault.`
`0x0807598e in YAML::Scanner::EnsureTokensInQueue() ()`

**ASAN:**
`ASAN:DEADLYSIGNAL`
`=================================================================
==2258==ERROR: AddressSanitizer: stack-overflow on address 0xbf338fa0 (pc 0x0814e423 bp 0xbf3393c8 sp 0xbf338f90 T0)`
`    #0 0x814e422  (/home/user/yaml-cpp/build/util/parse+0x814e422)`
`    #1 0x816fd48  (/home/user/yaml-cpp/build/util/parse+0x816fd48)`
`    #2 0x8190a72  (/home/user/yaml-cpp/build/util/parse+0x8190a72)`
`    #3 0x81d5061  (/home/user/yaml-cpp/build/util/parse+0x81d5061)`
`    #4 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`    #5 0x81d511a  (/home/user/yaml-cpp/build/util/parse+0x81d511a)`
`    #6 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`    #7 0x81d511a  (/home/user/yaml-cpp/build/util/parse+0x81d511a)`
`    #8 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`    #9 0x81d511a  (/home/user/yaml-cpp/build/util/parse+0x81d511a)`
`    #10 0x81deb85  (/home/user/yaml-cpp/build/util/parse+0x81deb85)`
`    #11 0x81d511a  (/home/user/yaml-cpp/build/util/parse+0x81d511a)`
`    ......                                                                                                         `
`    #245 0x81d511a  (/home/eack/yaml-cpp/build/util/parse+0x81d511a)`
`    #246 0x81deb85  (/home/eack/yaml-cpp/build/util/parse+0x81deb85)`
`    #247 0x81d511a  (/home/eack/yaml-cpp/build/util/parse+0x81d511a)`
`    #248 0x81deb85  (/home/eack/yaml-cpp/build/util/parse+0x81deb85)`
`    #249 0x81d511a  (/home/eack/yaml-cpp/build/util/parse+0x81d511a)`
`    #250 0x81deb85  (/home/eack/yaml-cpp/build/util/parse+0x81deb85)`
`SUMMARY: AddressSanitizer: stack-overflow (/home/eack/yaml-cpp/build/util/parse+0x814e422) 
==2258==ABORTING`

### LINKED ISSUE #660: Stack Overflow in SingleDocParser::HandleFlowSequence(YAML::EventHandler&) ()
Stack Overflow in SingleDocParser::HandleFlowSequence(YAML::EventHandler&) ()

position：[code](https://github.com/jbeder/yaml-cpp/blob/b87c76a2ef4f974e2739a2d410ee5e36122cfc00/src/singledocparser.cpp#L183)

[poc](https://github.com/jbeder/yaml-cpp/files/2754280/crash0.zip)

To reproduce: `./parse < crash0`

gdb:
`Program received signal SIGSEGV, Segmentation fault.`
`0x0807e61d in YAML::SingleDocParser::HandleFlowSequence(YAML::EventHandler&) ()`

ASAN:

    ASAN:DEADLYSIGNAL
    =================================================================
    ==9579==ERROR: AddressSanitizer: stack-overflow on address 0xbf6a8fc0 (pc 0x0814e3e3 bp 0xbf6a93e8 sp 0xbf6a8fb0 T0)
        #0 0x814e3e2  (/home/eack/yaml-cpp/build2/util/parse+0x814e3e2)
        #1 0x816fde8  (/home/eack/yaml-cpp/build2/util/parse+0x816fde8)
        #2 0x818f283  (/home/eack/yaml-cpp/build2/util/parse+0x818f283)
        #3 0x81d41b8  (/home/eack/yaml-cpp/build2/util/parse+0x81d41b8)
        #4 0x81d966a  (/home/eack/yaml-cpp/build2/util/parse+0x81d966a)
        #5 0x81d4528  (/home/eack/yaml-cpp/build2/util/parse+0x81d4528)
        #6 0x81dfdc9  (/home/eack/yaml-cpp/build2/util/parse+0x81dfdc9)
        #7 0x81d47fc  (/home/eack/yaml-cpp/build2/util/parse+0x81d47fc)
        #8 0x81d966a  (/home/eack/yaml-cpp/build2/util/parse+0x81d966a)
        #9 0x81d4528  (/home/eack/yaml-cpp/build2/util/parse+0x81d4528)
        #10 0x81dfdc9  (/home/eack/yaml-cpp/build2/util/parse+0x81dfdc9)
        #11 0x81d47fc  (/home/eack/yaml-cpp/build2/util/parse+0x81d47fc)
        #12 0x81d966a  (/home/eack/yaml-cpp/build2/util/parse+0x81d966a)
        #13 0x81d4528  (/home/eack/yaml-cpp/build2/util/parse+0x81d4528)
        #14 0x81dfdc9  (/home/eack/yaml-cpp/build2/util/parse+0x81dfdc9)
        #15 0x81d47fc  (/home/eack/yaml-cpp/build2/util/parse+0x81d47fc)
        ......
        #243 0x81d47fc  (/home/eack/yaml-cpp/build2/util/parse+0x81d47fc)
        #244 0x81d966a  (/home/eack/yaml-cpp/build2/util/parse+0x81d966a)
        #245 0x81d4528  (/home/eack/yaml-cpp/build2/util/parse+0x81d4528)
        #246 0x81dfdc9  (/home/eack/yaml-cpp/build2/util/parse+0x81dfdc9)
        #247 0x81d47fc  (/home/eack/yaml-cpp/build2/util/parse+0x81d47fc)
        #248 0x81d966a  (/home/eack/yaml-cpp/build2/util/parse+0x81d966a)
        #249 0x81d4528  (/home/eack/yaml-cpp/build2/util/parse+0x81d4528)
        #250 0x81dfdc9  (/home/eack/yaml-cpp/build2/util/parse+0x81dfdc9)
    
    SUMMARY: AddressSanitizer: stack-overflow (/home/eack/yaml-cpp/build2/util/parse+0x814e3e2) 
    ==9579==ABORTING

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Fix stack overflow. (Fixes: #459/CVE-2017-5950)

The same solution as #489 rebased on current master and with a test.

Also fixes #655/CVE-2018-20573
Also fixes #654/CVE-2018-20574
Also fixes #660/CVE-2019-6285

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
@jbeder is there a reason the proposed fix (#489) for these CVEs hasn't been merged? Anything I could do to fix that?

Background: to get yaml-cpp into Ubuntu "main" archive I'm trying to address the CVEs.

File: include/yaml-cpp/depthguard.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * @brief The DeepRecursion class
+ *  An exception class which is thrown by DepthGuard. Ideally it should be
+ * a member of DepthGuard. However, DepthGuard is a templated class which means
+ * that any catch points would then need to know the template parameters. It is
+ * simpler for clients to not have to know at the catch point what was the
+ * maximum depth.
+ */
+class DeepRecursion : public ParserException {
+  int m_atDepth = 0;
```
Comment (jbeder) [inline include/yaml-cpp/depthguard.h:23]: Put this in a separate "private:" block at the bottom, and name this variable m_depth.

File: include/yaml-cpp/depthguard.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+public:
+  // no custom dtor needed, but virtual dtor necessary to prevent slicing
```
Comment (jbeder) [inline include/yaml-cpp/depthguard.h:25]: Commnet is unnecessary.

File: include/yaml-cpp/depthguard.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  virtual ~DeepRecursion() = default;
+
+  // construct an exception explaining how deep you were
+  DeepRecursion(int at_depth, const Mark& mark_, const std::string& msg_);
+
+  // query how deep you were when the exception was thrown
```
Comment (jbeder) [inline include/yaml-cpp/depthguard.h:31]: Name this "depth()", and you can just inline the contents since it's trivial.

Comment (jbeder) [inline include/yaml-cpp/depthguard.h:31]: Comments should start with a capital letter and end with a period. Also in the third-person present tense, e.g.

"Returns the recursion depth when the exception was thrown."

File: include/yaml-cpp/depthguard.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * simpler for clients to not have to know at the catch point what was the
+ * maximum depth.
+ */
+class DeepRecursion : public ParserException {
+  int m_atDepth = 0;
+public:
+  // no custom dtor needed, but virtual dtor necessary to prevent slicing
+  virtual ~DeepRecursion() = default;
+
+  // construct an exception explaining how deep you were
```
Comment (jbeder) [inline include/yaml-cpp/depthguard.h:28]: Unnecessary comment.

File: include/yaml-cpp/depthguard.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  DeepRecursion(int at_depth, const Mark& mark_, const std::string& msg_);
```
Comment (jbeder) [inline include/yaml-cpp/depthguard.h:29]: Rename var as above.

File: include/yaml-cpp/depthguard.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * If the integer would be incremented past max_depth, then an exception is
+ * thrown. This is ideally geared toward guarding against deep recursion.
+ *
+ * @param max_depth
+ *  compile-time configurable maximum depth.
+ */
+template <int max_depth = 2000>
+class DepthGuard final /* final because non-virtual dtor */ {
+  int & m_depth;
```
Comment (jbeder) [inline include/yaml-cpp/depthguard.h:48]: Move var to private section below.

File: include/yaml-cpp/depthguard.h [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ * construction of DepthGuard and decrements the integer upon destruction.
+ *
+ * If the integer would be incremented past max_depth, then an exception is
+ * thrown. This is ideally geared toward guarding against deep recursion.
+ *
+ * @param max_depth
+ *  compile-time configurable maximum depth.
+ */
+template <int max_depth = 2000>
+class DepthGuard final /* final because non-virtual dtor */ {
```
Comment (jbeder) [inline include/yaml-cpp/depthguard.h:47]: Unnecessary comment.

File: include/yaml-cpp/depthguard.h [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  int & m_depth;
+public:
+  DepthGuard(int & depth_, const Mark& mark_, const std::string& msg_) : m_depth(depth_) {
+    ++m_depth;
+    if ( max_depth <= m_depth ) {
+        throw DeepRecursion{m_depth, mark_, msg_};
+    }
+  }
+
+  // DepthGuard is neither copyable nor moveable.
```
Comment (jbeder) [inline include/yaml-cpp/depthguard.h:57]: Unnecesssary comment.

File: src/singledocparser.h [hunk 9]
```diff
@@ -55,6 +57,8 @@ class SingleDocParser {
   anchor_t LookupAnchor(const Mark& mark, const std::string& name) const;
 
  private:
+  using DepthGuard = YAML::DepthGuard<2000>;
```
Comment (jbeder) [inline src/singledocparser.h:60]: Why do you need this using? You can just inline the call in the .cpp file.

File: src/singledocparser.h [hunk 10]
```diff
@@ -15,6 +15,8 @@
 
 namespace YAML {
 class CollectionStack;
+template <int> class DepthGuard; // depthguard.h
+class DeepRecursion; // an exception which may be thrown from excessive call stack recursion, see depthguard.h
```
Comment (jbeder) [inline src/singledocparser.h:19]: Why this forward declaration?

File: test/parser_test.cpp [hunk 11]
```diff
@@ -14,3 +16,39 @@ TEST(ParserTest, Empty) {
...
     EXPECT_FALSE(parser.HandleNextDocument(handler));
 }
+
+TEST(ParserTest, CVE_2017_5950) {
+    std::string excessive_recursion(16384, '[');
+    std::istringstream input{excessive_recursion};
+    Parser parser{input};
+
+    NiceMock<MockEventHandler> handler;
+    EXPECT_THROW(parser.HandleNextDocument(handler), YAML::ParserException);
```
Comment (jbeder) [inline test/parser_test.cpp:26]: Can you expect the specific exception?

File: test/parser_test.cpp [hunk 12]
```diff
@@ -14,3 +16,39 @@ TEST(ParserTest, Empty) {
...
+    std::string excessive_recursion(20535, '{');
+    std::istringstream input{excessive_recursion};
+    Parser parser{input};
+
+    NiceMock<MockEventHandler> handler;
+    EXPECT_THROW(parser.HandleNextDocument(handler), YAML::ParserException);
+}
+
+TEST(ParserTest, CVE_2018_20574) {
+    std::string excessive_recursion(21989, '{');
```
Comment (jbeder) [inline test/parser_test.cpp:39]: What's the difference between this and the one above, except the slightly different recursion depth?

File: include/yaml-cpp/depthguard.h (hunk 1 above)
Comment (AlanGriffiths) [reply include/yaml-cpp/depthguard.h]: Done

File: include/yaml-cpp/depthguard.h (hunk 2 above)
Comment (AlanGriffiths) [reply include/yaml-cpp/depthguard.h]: Deleted

File: include/yaml-cpp/depthguard.h (hunk 3 above)
Comment (AlanGriffiths) [reply include/yaml-cpp/depthguard.h]: Done

Comment (AlanGriffiths) [reply include/yaml-cpp/depthguard.h]: Done

File: include/yaml-cpp/depthguard.h (hunk 4 above)
Comment (AlanGriffiths) [reply include/yaml-cpp/depthguard.h]: Deleted

File: include/yaml-cpp/depthguard.h (hunk 8 above)
Comment (AlanGriffiths) [reply include/yaml-cpp/depthguard.h]: Deleted

File: include/yaml-cpp/depthguard.h (hunk 6 above)
Comment (AlanGriffiths) [reply include/yaml-cpp/depthguard.h]: Done

File: src/singledocparser.h (hunk 9 above)
Comment (AlanGriffiths) [reply src/singledocparser.h]: Reworked

File: src/singledocparser.h (hunk 10 above)
Comment (AlanGriffiths) [reply src/singledocparser.h]: Deleted

File: test/parser_test.cpp (hunk 11 above)
Comment (AlanGriffiths) [reply test/parser_test.cpp]: Sure

File: test/parser_test.cpp (hunk 12 above)
Comment (AlanGriffiths) [reply test/parser_test.cpp]: The CVEs differ by the recursion depths. (I know!)

File: include/yaml-cpp/depthguard.h (hunk 5 above)
Comment (AlanGriffiths) [reply include/yaml-cpp/depthguard.h:26]: Which var?

File: include/yaml-cpp/depthguard.h (hunk 7 above)
Comment (AlanGriffiths) [reply include/yaml-cpp/depthguard.h]: Deleted

File: include/yaml-cpp/depthguard.h (hunk 5 above)
Comment (jbeder) [reply include/yaml-cpp/depthguard.h:26]: at_depth

**Pull request merged**